File: day2/day2.ts
import fs from "fs";

interface Entry {
    game: number;
    red: number;
    green: number;
    blue: number;
}

const result = fs.readFileSync(process.argv[2], "utf-8")
    .split(/\r?\n/)
    .filter(line => line.length > 0)
    .map(line => {
        return parseLine(line)
    })
    .reduce((acc: number, current: Entry) => {
        return acc + current.red * current.green * current.blue;;
    }, 0)

console.log(result);

function parseLine(line: string) {
    const [game, data] = line.split(/\s*:\s*/);
    const result: Entry = {
        game: Number(game.substring(5)),
        red: 0,
        green: 0,
        blue: 0
    }
    data.split(/\s*;\s*/)
        .forEach(set => {
            set.split(/\s*,\s*/)
                .forEach(entry => {
                    const [count, color] = entry.split(" ");
                    if (isValidColor(color)) {
                        result[color] = Math.max(result[color], Number(count));
                    }
                });
            return result;
        });

    return result;
}

function isValidColor(color: string): color is keyof Entry {
    return color === "red" || color === "green" || color === "blue";
}

File: day2/day2.go
package adventofcode

import util "adventofcode/util"

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func Run(args []string) {   
	if len(args) == 0 {
		fmt.Println("A file must be specified")
		os.Exit(1)
	} 

	file, err := os.Open(args[0])
    util.Check(err)
    defer file.Close()

	scanner := bufio.NewScanner(file)
	result := 0
    for scanner.Scan() { 
		result += parseLine(scanner.Text())
    }

	util.Check(scanner.Err())

	fmt.Println(result)
}

func parseLine(line string) int {
	if len(line) == 0 {
		return 0
	}
	parts := strings.Split(line, ":")
	_, data := strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])

	result := map[string]int {
		"red": 0,
		"green": 0,
		"blue": 0,
	}

	for _, set := range strings.Split(data, ";") {
		for _, entry := range strings.Split(strings.TrimSpace(set), ",") {
			parts = strings.Split(strings.TrimSpace(entry), " ")
			count, err := strconv.Atoi(strings.TrimSpace(parts[0]))
			util.Check(err)
			color := strings.TrimSpace(parts[1])
			result[color] = util.Max(result[color], count)
		}
	}

    return result["red"] * result["green"] * result["blue"]
}
